Fix template footer click to focus canvas

jeryj opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Clicking the Template button in the footer breadcrumbs causes a focus loss to the body of the page. This PR sets focus to the editor canvas instead, making it similar to how the other buttons in footer breadcrumbs work.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
a11y. Focus management.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
When clicking. "Template" in the footer breadcrumbs, set focus on the editor canvas.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
- Select a block
- Click the root "Template" button in the footer
- Focus should be visible on the editor canvas

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->


https://github.com/WordPress/gutenberg/assets/967608/dcb0e4cb-2352-4ee0-aa40-99d04ce5af08

jeryj commented on packages/block-editor/src/components/block-breadcrumb/index.js:
```diff
@@ -60,7 +65,11 @@ function BlockBreadcrumb( { rootLabelText } ) {
+						onClick={ () => {
+							clearSelectedBlock();
+							// Focus the editor iframe.
+							blockRef.current?.closest( 'body' )?.focus();
```
@ellatrix Is there a better way to get the editor canvas ref?

andrewserong replied:
Would it be worth trying the `.editor-styles-wrapper` classname, since that gets attached to the `body` element?

ellatrix replied:
Not really. But maybe, when selection is cleared, focus should automatically be set to the writing flow element or something (by the writing flow component, or iframe component maybe)

ellatrix replied:
See also https://github.com/WordPress/gutenberg/pull/62196/files#r1630727302

andrewserong reviewed (commented):
It looks like attempting to focus `body` only works in the iframed editor, and not if the editor doesn't use an iframe. For example, in the post editor if custom fields are enabled in preferences.

https://github.com/WordPress/gutenberg/assets/14988353/88c896b1-a0ca-4ed5-af00-f90dbb427dff

In the above video, this PR seems to work nicely when iframed, but then when switching the custom fields on, there's still a focus loss when clicking on "Post". Is that a blocker? Now that I think of it, is there a tab stop for the canvas in the non-iframed editor 🤔

jeryj commented:
Thanks for the review, @andrewserong! Great catch on the non-iframed version too. I updated the focus styles to work for both versions of the editor.

andrewserong reviewed (commented):
Thanks for the update, this is feeling closer to me. I still noticed a couple of oddities in the non-iframed editor while running TT4 theme (I'm testing in Chrome on a Mac in case that's also part of it).

* When in fullscreen mode in the non-iframed editor, the outline is only visible to the top and to the right
* When not in fullscreen mode, the outline appears to be displaying inconsistently for me. Not sure if it's hidden behind other UI elements (e.g. if you scroll you can kind of see part of the outline disappearing beneath the custom fields)? I wondered if the non-iframed version might work better with absolute positioning instead of fixed 🤔

| Non-iframed fullscreen | Non-iframed not fullscreen |
| --- | --- |
| ![image](https://github.com/WordPress/gutenberg/assets/14988353/85600cce-f423-452a-81b2-030a3e26935b) | ![image](https://github.com/WordPress/gutenberg/assets/14988353/a648d890-fc18-476a-bbe8-b8926e69cb0e) |

Another thing I noticed while testing is that the outline is also visible for me when clicking on the editor canvas to deselect a block. Is that intentional? I wasn't sure if the scope of this PR was just about when clicking on the breadcrumbs in the interface footer, or if it should also apply with this interaction. It's a fairly prominent change, so just thought I'd raise it:

https://github.com/WordPress/gutenberg/assets/14988353/6300a1af-bf16-4986-a680-a98bbdd40018

jeryj commented:
> Another thing I noticed while testing is that the outline is also visible for me when clicking on the editor canvas to deselect a block. Is that intentional? 

It's not necessarily intentional, but the expected behavior since it's placing focus on the body element of the canvas. It focusable on trunk as well, but doesn't show you that focus is placed there.

I think these explorations and testing are really helpful. At this point, I think we should solve where focus goes on https://github.com/WordPress/gutenberg/pull/62196, as this one should get the same treatment.

ellatrix commented:
Should this be backported for WP 6.6? If so, please add the `Backport to WP Beta/RC` label. Thanks!

jeryj committed:
Redo CSS for editor canvas outline.
- moves selected-block-outline to base styles css so it can be used anywhere
- renames the mixin to match mixin naming convention

jeryj committed:
Add a focusable wrapper for the block canvas

jeryj committed:
Clear focus from template click to the focusable wrapper

jeryj committed:
Fix wrapper selector

kevin940726 commented on packages/block-editor/src/components/block-breadcrumb/index.js:
```diff
@@ -60,7 +65,13 @@ function BlockBreadcrumb( { rootLabelText } ) {
+							clearSelectedBlock();
+							// Focus the editor iframe.
+							blockRef.current
+								?.closest( '[aria-label="Block canvas"]' )
```
Any reason why we don't just create a ref for `FocusableWrapper` and pass it down via context?

jeryj replied:
That's a good idea -- I think we're going to try to focus the editor region instead though.

scruffian replied:
I added a ref to do this. The downside is that the ref needs to be passed through props.

scruffian reviewed (commented):
In my testing the focus isn't visible unless I use my keyboard to select the breadcrumb items. Not sure if that's deliberate?

scruffian committed:
Focus the wrapper with a ref

jeryj commented:
Closing in favor of https://github.com/WordPress/gutenberg/pull/62595
